sys-utils/libcmt/src/io-mock.c

```diff
@@ -0,0 +1,302 @@
+#include <string.h>
+
+#define CMT_TX_BUF_MAX_LENGTH (2UL * 1024 * 1024) // 2MB
+#define CMT_RX_BUF_MAX_LENGTH (2UL * 1024 * 1024) // 2MB
+
+static int read_whole_file(const char *name, size_t max, void *data, size_t *length);
+static int write_whole_file(const char *name, size_t length, const void *data);
+
+int cmt_io_init(cmt_io_driver_t *_me) {
+    if (!_me) {
+        return -EINVAL;
+    }
+    cmt_io_driver_mock_t *me = &_me->mock;
+
+    cmt_buf_init(me->tx, CMT_TX_BUF_MAX_LENGTH, malloc(CMT_TX_BUF_MAX_LENGTH));
+    cmt_buf_init(me->rx, CMT_RX_BUF_MAX_LENGTH, malloc(CMT_RX_BUF_MAX_LENGTH));
+
+    if (!me->tx->begin || !me->rx->begin) {
+        free(me->tx->begin);
+        free(me->rx->begin);
+        return -ENOMEM;
+    }
+
+    char *inputs = getenv("CMT_INPUTS");
+    if (inputs) {
+        cmt_buf_init(&me->inputs_left, strlen(inputs), inputs);
+    } else {
+        cmt_buf_init(&me->inputs_left, 0, "");
+    }
+
+    // in case the user writes something before loading any input
+    strcpy(me->input_filename, "none");
+    strcpy(me->input_fileext, ".bin");
+
+    me->input_seq = 0;
+    me->output_seq = 0;
+    me->report_seq = 0;
+    me->exception_seq = 0;
+
+    return 0;
+}
+
+void cmt_io_fini(cmt_io_driver_t *_me) {
+    if (!_me) {
+        return;
+    }
+    cmt_io_driver_mock_t *me = &_me->mock;
+
+    free(me->tx->begin);
+    free(me->rx->begin);
+}
+
+cmt_buf_t cmt_io_get_tx(cmt_io_driver_t *me) {
+    cmt_buf_t empty = {NULL, NULL};
+    if (!me) {
+        return empty;
+    }
+    return *me->mock.tx;
+}
+
+cmt_buf_t cmt_io_get_rx(cmt_io_driver_t *me) {
+    cmt_buf_t empty = {NULL, NULL};
+    if (!me) {
+        return empty;
+    }
+    return *me->mock.rx;
+}
+
+static int load_next_input(cmt_io_driver_mock_t *me, struct cmt_io_yield *rr) {
+    cmt_buf_t current_input;
+    char filepath[128] = {0};
+    if (!cmt_buf_split_by_comma(&current_input, &me->inputs_left)) {
+        return -EINVAL;
+    }
+    // NOLINTNEXTLINE(cert-err34-c,clang-analyzer-security.insecureAPI.DeprecatedOrUnsafeBufferHandling)
+    if (sscanf((char *) current_input.begin, "%d:%127[^,]", &me->input_type, filepath) != 2) {
+        return -EINVAL;
+    }
+
+    size_t file_length = 0;
+    int rc = read_whole_file(filepath, CMT_RX_BUF_MAX_LENGTH, me->rx->begin, &file_length);
+    if (rc) {
+        (void) fprintf(stderr, "failed to load \"%s\". %s\n", filepath, strerror(-rc));
+        return rc;
+    }
+    // NOLINTNEXTLINE(clang-analyzer-security.insecureAPI.DeprecatedOrUnsafeBufferHandling)
+    if (sscanf(filepath, " %127[^.]%15s", me->input_filename, me->input_fileext) != 2) {
+        (void) fprintf(stderr, "failed to parse filename: \"%s\"\n", filepath);
+        return -EINVAL;
+    }
+
+    assert(file_length <= INT32_MAX);
+    rr->reason = me->input_type;
+    rr->data = file_length;
+
+    me->output_seq = 0;
+    me->report_seq = 0;
+    me->exception_seq = 0;
+    me->rx->end = me->rx->begin + file_length;
```

Comment:
`tx` and `rx` buffers are designed to mirror the kernel mmap buffers and thus be read only throughout the library use.
The length is already conveyed in the `data` field of yield.